### PR TITLE
Implement a Vault plugin

### PR DESCRIPTION
**Pull request by ramnes**

### CODE REVIEW

**Comment by rambobinator** [comment]
Hi,

I'm not familiar with vault, but we agree that we don't handle any authentication here? or is it managed on hvac side ?

**Comment by ramnes** [comment]
It's similar to Kubernetes, you login with the Vault CLI and then HVAC can use the token.

File: envfrom/plugins/vault.py [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+    def __init__(self, path, addr=None):
+        self.paths = path
+        if addr is None:
+            addr = os.environ.get("VAULT_ADDR", "https://127.0.0.1:8200")
+        self.client = hvac.Client(url=addr)
+
+        if "KUBERNETES_PORT" in os.environ:
+            f = open('/var/run/secrets/kubernetes.io/serviceaccount/token')
+            jwt = f.read()
+            self.client.auth_kubernetes("default", jwt)
```
Comment (ramnes) [inline envfrom/plugins/vault.py:31]: "default" is hardcoded right now, it should be made dynamic before merging

File: envfrom/plugins/vault.py [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+            if value.startswith("vault:"):
+                env[key] = self.get_secret(value[6:])
+
+        for path in self.paths:
+            secret = self.get_secret(path)
+            if isinstance(secret, dict):
+                for key, value in secret.items():
+                    env[key.upper()] = value
+            else:
+                env[key.upper()] = secret
```
Comment (ramnes) [inline envfrom/plugins/vault.py:63]: `key` doesn't exist, we need to get it after the `#` or specify it in the same fashion than the CLI plugin.

File: envfrom/plugins/vault.py [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+        data = response["data"]["data"]
+        if key:
+            return data[key]
+        return data
+
+    def process(self):
+        env = {}
+
+        for key, value in os.environ.items():
+            if value.startswith("vault:"):
```
Comment (ramnes) [inline envfrom/plugins/vault.py:54]: Maybe raise an error if no key was specified with `#`?

**Pull request merged**